src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -470,86 +461,62 @@ internal unsafe SocketError DoOperationReceiveFromMultiBuffer(SafeSocketHandle h
-                {
-                    _controlBufferGCHandle.Free();
-                }
-                _controlBuffer = new byte[sizeof(Interop.Winsock.ControlData)];
+                _controlBufferPinned = GC.AllocateUninitializedArray<byte>(sizeof(Interop.Winsock.ControlData), pinned: true);
             }
-            else if (ipv6 && (_controlBuffer == null || _controlBuffer.Length != sizeof(Interop.Winsock.ControlDataIPv6)))
+            else if (ipv6 && (_controlBufferPinned == null || _controlBufferPinned.Length != sizeof(Interop.Winsock.ControlDataIPv6)))
             {
-                if (_controlBufferGCHandle.IsAllocated)
-                {
-                    _controlBufferGCHandle.Free();
-                }
-                _controlBuffer = new byte[sizeof(Interop.Winsock.ControlDataIPv6)];
+                _controlBufferPinned = GC.AllocateUninitializedArray<byte>(sizeof(Interop.Winsock.ControlDataIPv6), pinned: true);
             }
 
             // If single buffer we need a single element WSABuffer.
             WSABuffer[] wsaRecvMsgWSABufferArray;
             uint wsaRecvMsgWSABufferCount;
             if (_bufferList == null)
             {
-                if (_wsaRecvMsgWSABufferArray == null)
+                if (_wsaRecvMsgWSABufferArrayPinned == null)
                 {
-                    _wsaRecvMsgWSABufferArray = new WSABuffer[1];
+                    _wsaRecvMsgWSABufferArrayPinned = GC.AllocateUninitializedArray<WSABuffer>(1, pinned: true);
                 }
 
                 Debug.Assert(_singleBufferHandleState == SingleBufferHandleState.None);
                 _singleBufferHandle = _buffer.Pin();
                 _singleBufferHandleState = SingleBufferHandleState.Set;
 
-                _wsaRecvMsgWSABufferArray[0].Pointer = (IntPtr)_singleBufferHandle.Pointer;
-                _wsaRecvMsgWSABufferArray[0].Length = _count;
-                wsaRecvMsgWSABufferArray = _wsaRecvMsgWSABufferArray;
+                _wsaRecvMsgWSABufferArrayPinned[0].Pointer = (IntPtr)_singleBufferHandle.Pointer;
+                _wsaRecvMsgWSABufferArrayPinned[0].Length = _count;
+                wsaRecvMsgWSABufferArray = _wsaRecvMsgWSABufferArrayPinned;
                 wsaRecvMsgWSABufferCount = 1;
             }
             else
             {
                 // Use the multi-buffer WSABuffer.
-                wsaRecvMsgWSABufferArray = _wsaBufferArray!;
+                wsaRecvMsgWSABufferArray = _wsaBufferArrayPinned!;
                 wsaRecvMsgWSABufferCount = (uint)_bufferListInternal!.Count;
             }
 
-            // Ensure the array is pinned.
-            Debug.Assert(!_wsaRecvMsgWSABufferArrayGCHandle.IsAllocated || _wsaRecvMsgWSABufferArrayGCHandle.Target == wsaRecvMsgWSABufferArray);
-            if (!_wsaRecvMsgWSABufferArrayGCHandle.IsAllocated)
-            {
-                _wsaRecvMsgWSABufferArrayGCHandle = GCHandle.Alloc(wsaRecvMsgWSABufferArray, GCHandleType.Pinned);
-            }
```

Comment:
GCHandle.Alloc is not only keeping the array pinned, it's also keeping it rooted, whereas the GC.Allocate* calls are just creating pinned arrays but not keeping them rooted.  In all of these uses where we've replaced the GCHandle.Alloc calls, have we proved to ourselves sufficiently that the array won't become garbage in the interim?

Reply:
When we access something from both managed and native code, we typically:
- keep both a handle and a field for the data, or 
- keep just a field, while pinning the buffer temporarily via a handle or by referring to it from Overlapped.

It is possible to fetch the object from the handle via `.Target`. I do not think this code relied on that, except in asserts - to make sure the handle matches the buffer.